projects/plugins/jetpack/_inc/lib/widgets.php

```diff
@@ -42,10 +42,10 @@ public static function get_sidebars_widgets() {
 	 *  )
 	 * )
 	 *
-	 * @param string|integer   $position The position of the widget in its sidebar.
-	 * @param string           $widget_id The widget's id (eg: 'text-3').
-	 * @param string           $sidebar The widget's sidebar id (eg: 'sidebar-1').
-	 * @param array (Optional) $settings The settings for the widget.
+	 * @param string|integer $position The position of the widget in its sidebar.
+	 * @param string         $widget_id The widget's id (eg: 'text-3').
+	 * @param string         $sidebar The widget's sidebar id (eg: 'sidebar-1').
+	 * @param array|null     $settings The settings for the widget.
```

Comment:
Should it be something like this?

```suggestion
	 * @param ?array|null     $settings The settings for the widget.
```

Reply:
That would be redundant, `?array` is the same as `array|null` so that would effectively be `array|null|null`.

It could be
```suggestion
	 * @param ?array         $settings The settings for the widget.
```
instead though. I don't think there's any difference between the two, and I don't think we currently have a preference.

Reply:
I guess I'm unclear if the PHPDoc defines what is passed to the function call vs. what is evaluated on the call. I was reading `?array` as an optionally-defined param , but it sounds instead like it's an optionally-null param, taking into account default params specified in the function definition.

Reply:
PHP doesn't have an "undefined" separate from `null`.

`?array` is literally just a shorthand for `array|null`. In part it exists because PHP's own type declarations added nullable types (i.e `?array` syntax) in 7.1 but didn't get general intersections (i.e. `array|null`, as well as stuff like `string|array`) until 8.1. The `?array` syntax replaces the older and now-deprecated `array $var = null` syntax for indicating that a parameter is nullable, which had some shortcomings.

I don't think there's a convention for "this parameter may default to null, but when calling it you should never pass null explicitly", if that's what you're asking about?

Reply:
> PHP doesn't have an "undefined" separate from `null`.

I guess it depends how you interpret this. Having a `null` value is not the same as being undefined; PHP has an "undefined" vs. "defined" var state, and any undefined var's value is `null`.

```
<?php
var_dump($undefined); // gives a warning
$null = null;
var_dump($null); // no warning
```

But in the context of types, this is true enough (there's no "undefined" type), and it's also true that within the function body, the function's params cannot be undefined.

> `?array` is literally just a shorthand for `array|null`

This is good to know; I was reading it as "optional param", which is a bit different. 😓